src/arch/Sound/DSoundHelpers.cpp

```diff
@@ -366,58 +362,54 @@ void DSoundBuf::CheckWriteahead( int iCursorStart, int iCursorEnd )
 /* Figure out if we've underrun, and act if appropriate. */
 void DSoundBuf::CheckUnderrun( int iCursorStart, int iCursorEnd )
 {
-	/* If the buffer is full, we can't be underrunning. */
-	if( m_iBufferBytesFilled >= m_iBufferSize )
-		return;
+	// Store the buffer size in a variable to avoid multiple calls
+	int storedSizeOfBuffer = m_iBufferSize;
 
-	/* If nothing is expected to be filled, we can't underrun. */
-	if( iCursorStart == iCursorEnd )
-		return;
-
-	/* If we're already in a recovering-from-underrun state, stop. */
-	if( m_iExtraWriteahead )
+	// If the buffer is full or nothing is expected to be filled, stop.
+	if( m_iBufferBytesFilled >= storedSizeOfBuffer || iCursorStart == iCursorEnd || m_iExtraWriteahead )
```

Comment:
I think original is fine. The comments help, and I don't think it hurts readability.